### PR TITLE
win midi: Compatibility update

### PR DESCRIPTION
**Pull request by ceski-1**
See discussion here: https://github.com/fabiangreffrath/woof/pull/1053

### CODE REVIEW

File: src/i_winmusic.c [hunk 1]
```diff
@@ -230,73 +230,77 @@ static void StreamOut(void)
     }
 }
 
-static void SendShortMsg(int time, int status, int channel, int param1, int param2)
+static void SendShortMsg(unsigned int delta_time, byte status, byte channel,
```
Comment (turol) [inline src/i_winmusic.c:241]: Are you sure this is safe? `MAKE_EVT` uses bitshifts on these without upcast. I don't remember the C implicit casting rules in that case.

File: src/i_winmusic.c [hunk 2]
```diff
@@ -230,73 +230,77 @@ static void StreamOut(void)
...
     WriteBuffer((byte *)&native_event, sizeof(native_event_t));
 }
 
-static void SendManualVolumeMsg(int time, int channel, int volume)
+static void SendManualVolumeMsg(unsigned int delta_time, byte channel,
+                                byte volume)
 {
-    int scaled_volume = volume * volume_factor + 0.5f;
-    SendShortMsg(time, MIDI_EVENT_CONTROLLER, channel,
+    const byte scaled_volume = (byte)(volume * volume_factor + 0.5f) & 0x7F;
```
Comment (turol) [inline src/i_winmusic.c]: What happens if this gets truncated? Should it saturate to 0x7F instead?

File: src/midifallback.c [hunk 3]
```diff
@@ -18,6 +18,7 @@
 #ifdef _WIN32
 
 #include "doomtype.h"
+#include <stdio.h>
```
Comment (turol) [inline src/midifallback.c:21]: Either before or after the all the existing ones so you're not mixing up includes with `" "` and `< >`.

File: src/midifallback.c [hunk 4]
```diff
@@ -111,6 +112,11 @@ static boolean GetProgramFallback(byte idx, byte program,
             // Fall to capital when no instrument has (successfully)
             // selected this variation or if the variation is above 63.
             fallback->value = 0;
+
+            fprintf(stderr,
```
Comment (turol) [inline src/midifallback.c:116]: How much spam can these generate? Should they be limited to once only? Or once per .midi?

File: src/m_misc.h [hunk 5]
```diff
@@ -28,6 +28,7 @@
 
 #ifdef _WIN32
 wchar_t *M_ConvertUtf8ToWide(const char *str);
+char *ConvertWideToUtf8(const wchar_t *wstr);
```
Comment (turol) [inline src/m_misc.h:31]: Since this is no longer static, document that the return value must be freed by caller.

Moving this should be in its own commit just in case this PR needs to be reverted.

File: src/i_winmusic.c [hunk 6]
```diff
@@ -43,6 +50,7 @@ enum
 };
 
 char *winmm_midi_device = NULL;
+int winmm_complevel = COMP_STANDARD;
```
Comment (turol) [inline src/i_winmusic.c:53]: Should this default to `COMP_VANILLA`? This is Chocolate Doom after all.

File: src/i_winmusic.c [hunk 7]
```diff
@@ -153,7 +153,7 @@ static void MidiError(const char *prefix, DWORD dwError)
     wchar_t werror[MAXERRORLENGTH];
     MMRESULT mmr;
 
-    mmr = midiOutGetErrorTextW(dwError, (LPWSTR)werror, MAXERRORLENGTH);
+    mmr = midiOutGetErrorTextW(dwError, (LPWSTR) werror, MAXERRORLENGTH);
```
Comment (turol) [inline src/i_winmusic.c:156]: Are these fixes to existing lines or your new commits? In the latter case they should be rebased back to the appropriate commit.

**Review by turol — CHANGES_REQUESTED** [review]

File: src/i_winmusic.c (hunk 1 above)
Comment (ceski-1) [reply src/i_winmusic.c:241]: The bytes are promoted to ints before bitshifting.

File: src/i_winmusic.c (hunk 2 above)
Comment (ceski-1) [reply src/i_winmusic.c]: See https://github.com/chocolate-doom/chocolate-doom/pull/1598/commits/1b0c89ea5ec365b6230b707562754348190ae8ba

File: src/midifallback.c (hunk 4 above)
Comment (ceski-1) [reply src/midifallback.c]: See above.

File: src/m_misc.h (hunk 5 above)
Comment (ceski-1) [reply src/m_misc.h]: Separated into https://github.com/chocolate-doom/chocolate-doom/pull/1598/commits/db56db1e799867bd78bd8716a0680322f42fe72a and https://github.com/chocolate-doom/chocolate-doom/pull/1598/commits/d126056cd725759eaf925d88c4e871743a6e8109

File: src/i_winmusic.c (hunk 6 above)
Comment (ceski-1) [reply src/i_winmusic.c]: Fixed by https://github.com/chocolate-doom/chocolate-doom/pull/1598/commits/48868398eb0bd02cc16ba2d28828ffafbc251876

File: src/i_winmusic.c (hunk 1 above)
Comment (ceski-1) [reply src/i_winmusic.c:242]: Fixed by https://github.com/chocolate-doom/chocolate-doom/pull/1598/commits/74cd516fa5c9a90e73d297d41c30f62b1f2da48f

**Review by fabiangreffrath — APPROVED** [review]

**Review by turol — APPROVED** [review]
